### DISCUSSION TITLE
Chunks commons bundle big because of ellipses, bn.js etc

### DISCUSSION THREAD

**Discussion by djeka07** [vercel/next.js#13631] [Help] [answered]
Hi all,

My commons chunk file has become big, stat size 3.07 mb and gzipped size is almost 400kb. I noticed when i analyzed it that elliptic, browserify-sign and bn.js takes alot of size, I don't know why the are getting bundled. Google gave me an answer about using crypto but i dont use that. It seems like it comes from node-libs-browser. Anyone got any ideas how to find where they are getting bundled in?

Thanks in advance!
![commons](https://user-images.githubusercontent.com/7669015/83417266-254f0380-a422-11ea-8495-54d0b05846e6.png)

**Comment by Timer** [comment, marked as answer]
This would likely be due to a dependency being used in your `pages/_app.js` file.

**Reply by djeka07** [reply]
Hi,

Thats what i thought as well. Removed _app.js and _document.js and they are still getting bundled. Any ideas how or where to look and tips why these libraries are in my client chunk?

<img width="1781" alt="Screenshot 2020-06-03 at 12 05 05" src="https://user-images.githubusercontent.com/7669015/83624117-81d13080-a592-11ea-9318-982bb787f454.png">

**Comment by robinwander** [comment]
@djeka07 I was wondering if you found out anything else about this. I'm noticing the same problem and I am a little confused. It looks like they are dependencies of webpack via next but I don't really understand how they would end up as dependencies of the client bundle.

@Timer could you elaborate on what dependencies might be the common culprit for this? I've removed _app and _document and still seeing these webpack dependencies in the common bundle.

**Reply by djeka07** [reply]
Hi!

Sorry for the late answer. Yes i actually found it after a while. I am using redux with persist and i was sharing code that was running from node and the client. Was this row "const { CookieStorage, NodeCookiesWrapper } = require('redux-persist-cookie-storage');" that was my problem, "NodeCookiesWrapper" was the one bundling in everything. Moved it outside to the calling files and that solved it.

Have you found anything yourself?

//André

**Reply by Nico924** [reply]
André, thank you very much! 
I had the exact same issue with the same package and without your comment I would never have found that just splitting my client and server store configuration was the solution.
Again thank you

Cheers

**Comment by jatinvmehta** [comment]
Any update on this ?

**Comment by switz** [comment]
This is due to a node.js targeted module that you're importing in the browser, for example [it could be] jsonwebtoken. Browserify bundles node.js internal modules for the browser that are huge. Instead use a module that is intended for the browser, like jwt-decode.

**Reply by wyantb** [reply]
Thank you, @switz , for the jsonwebtoken mention! It's probably a red herring for a lot of people, but in my case this was our exact silent (and hidden) dependency adding at least 162kb of gzipped to every page load. 

Indeed nextjs won't bundle these kinds of crypto polyfills in the browser modules themselves -- they could very well include them as a dep for their server code but leave it unreferenced by client code. I could see that after removing our jsonwebtoken import.

**Reply by luiznasciment0** [reply]
Thank you, @switz ! But what if I need to use JWT in a server-side function? I mean, I'm using getInitialProps and inside of it I have a function that I've imported from a module where I use JWT to decode a cookie

**Reply by switz** [reply]
If you want to just decode it, use jwt-decode. If you want to verify and decode it, I think you can require('jsonwebtoken') inside the `getServerSideProps` function (and maybe even `getInitialProps`?) and it won't put it in your client side bundle. You'll have to do some testing w/ various configuration to confirm that though, I don't know off the top of my head.

**Reply by luiznasciment0** [reply]
Awesome suggestions! Thanks a lot for the quick answer <3

**Reply by luiznasciment0** [reply]
Your suggestion to use jwt-decode + require it inside getServerSideProps solved my problem like a charm. Thanks!!

**Reply by Allyday** [reply]
Thank you!! For us it was the `web3` package in a client file. After removing it, the chunk size was reduced by a whopping **3.3MB**!!

**Comment by vuamitom** [comment]
I think it is because of Next dependency on crypto-browsersify, which in turn depends on browsersify-sign, which depends on bn.js

**Reply by switz** [reply]
Unless something changed in a recent version, next itself wouldn't bundle it; it's probably an external module that you're using.

**Reply by vuamitom** [reply]
I see that crypto-browserify is specified as next's dependency here: [https://www.npmjs.com/package/next](https://www.npmjs.com/package/next)

**Reply by switz** [reply]
https://github.com/vercel/next.js/commit/843d58425b65fc00cfc1b416128fdd3343a3b083

Shouldn’t be included unless you’re importing a node module (or an external
module that depends on a node module)

**Comment by kennandavison** [comment]
Hey @djeka07, did you ever figure this out? I see the same seemingly next-affiliated packages being bundled as well and a bit stumped.

**Reply by djeka07** [reply]
Hi! 

Sorry for the late answer. Yes i actually found it after a while. I am using redux with persist and i was sharing code that was running from node and the client. Was this row "const { CookieStorage, NodeCookiesWrapper } = require('redux-persist-cookie-storage');" that was my problem, "NodeCookiesWrapper" was the one bundling in everything. Moved it outside to the calling files and that solved it. 

Have you found anything yourself? 

//André

**Comment by roelvan** [comment]
For reference: I accidentally used `import crypto from 'crypto';` on the client, made my bundle huge (node_modules bn.js)!

**Reply by bsplosion** [reply]
This was it for me as well!  We'd used `import { randomBytes } from 'crypto';` which was triggering the entire `browserify` dependency stack including all of the `bn.js` dependencies.

I have a feeling the real answer to this is: if you use `crypto` _anywhere_ in your client side, NextJS will bundle a _lot_ of extra dependencies and balloon your package.  Find alternatives if possible.

**Edit:** also, if you're looking for suitable replacements, check out the browserify org for a variety of standalone options:
https://github.com/orgs/crypto-browserify/repositories?type=all

**Comment by danboyle8637** [comment]
I too imported a node library in my client for use in a utility function that I used in my serverless functions. When I moved that into my functions... all instances of bn.js were gone. It seems any node only library on the client might make this happen.

**Comment by DouglasGabr** [comment]
Had the same issue, for me it was because I was using `jsonwebtoken` to decode a token. Switched it for `jwt-decode` and problem solved

**Comment by fnavarijo** [comment]
Really appreciate this thread. For me, it was the `url` node package. It was adding almost 200kB.

Thanks for sharing!

**Comment by dm-rybalchenko** [comment]
Thanks for the thread, for us it was `imgproxy`.
